### PR TITLE
Lazily Resume/Restore/Create filecoin keyring 

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/20707
Resolves https://github.com/brave/brave-browser/issues/20675

- Pass filecoin flag to UI and show filecoin UI if enabled
- Resume/Restore/Create filecoin keyring if user enabled experimental flag. The keyring will be lazily created when user decides to import or create his first account.
- Import filecoin accounts from Ledger or private keys

- Create filecoin accounts in wallet:

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Restore/Resume/Create keyrings in wallet and make sure it is not broken same when enabling filecoin experimental flag before and after default keyring creation

### CODE REVIEW

File: components/brave_wallet/browser/keyring_service.cc [hunk 1]
```diff
@@ -521,11 +521,12 @@ void KeyringService::RequestUnlock() {
 
 HDKeyring* KeyringService::ResumeKeyring(const std::string& keyring_id,
                                          const std::string& password) {
-  if (!CreateEncryptorForKeyring(password, keyring_id)) {
+  if (!encryptor_ && !CreateEncryptorForKeyring(password, keyring_id)) {
```
Comment (darkdh) [inline components/brave_wallet/browser/keyring_service.cc:524]: Encryptor and its salt is designed to be used per keyring, so different keyring can't share the same encryptor
We can share same password but not encryptor, use map to store encryptor per keyring keyed by keyring id

File: components/brave_wallet/browser/keyring_service.cc [hunk 2]
```diff
@@ -521,11 +521,12 @@ void KeyringService::RequestUnlock() {
...
     return nullptr;
   }
 
-  const std::string mnemonic = GetMnemonicForKeyringImpl(keyring_id);
+  const std::string mnemonic =
+      GetMnemonicForKeyringImpl(mojom::kDefaultKeyringId);
```
Comment (darkdh) [inline components/brave_wallet/browser/keyring_service.cc:529]: as discussed on Slack, we need to store different encrypted mnemonic per keyring because we have separate encryptor for each keyring even though we share the same plaintext mnemonic among keyrings.

File: components/brave_wallet/browser/keyring_service.cc [hunk 3]
```diff
@@ -1314,16 +1328,18 @@ void KeyringService::IsLocked(IsLockedCallback callback) {
...
+                           bool notify_observer) {
   StopAutoLockTimer();
-  encryptor_.reset();
-  keyrings_.clear();
-
-  ClearKeyringServiceProfilePrefs(prefs_);
+  encryptors_.erase(keyring_id);
+  keyrings_.erase(keyring_id);
+  if (keyring_id == mojom::kDefaultKeyringId) {
+    ClearKeyringServiceProfilePrefs(prefs_);
```
Comment (darkdh) [inline components/brave_wallet/browser/keyring_service.cc:1337]: Reset with keyring_id is problematic because this will clear `kBraveWalletKeyrings` when keyring is default, so we also need to clear all encryptors_ and keyrings_ because other keyring's prefs are cleared too. That being said, default keyring must exists for other types of keyrings. 
I think it is better to leave it unchanged in this PR and we can open a separate issue to address it.

**Review by darkdh — COMMENTED** [review]
We also need to test the same mnemonic is stored as different encrypted value in different keyrings. And test Lock and Unlock with multiple keyrings created

File: components/brave_wallet/browser/keyring_service.cc (hunk 3 above)
Comment (darkdh) [reply components/brave_wallet/browser/keyring_service.cc:1337]: issue opened here
https://github.com/brave/brave-browser/issues/20731

File: browser/brave_wallet/keyring_service_unittest.cc [hunk 4]
```diff
@@ -477,9 +477,9 @@ TEST_F(KeyringServiceUnitTest, CreateEncryptorForKeyring) {
   {
     KeyringService service(GetPrefs());
     EXPECT_TRUE(service.CreateEncryptorForKeyring("123", "default"));
-    EXPECT_NE(service.encryptor_, nullptr);
+    EXPECT_NE(service.encryptors_.at("default"), nullptr);
     EXPECT_TRUE(service.CreateEncryptorForKeyring("456", "keyring2"));
-    EXPECT_NE(service.encryptor_, nullptr);
+    EXPECT_NE(service.encryptors_.at("keyring2"), nullptr);
```
Comment (darkdh) [inline browser/brave_wallet/keyring_service_unittest.cc:482]: need to check encryptor of default is not the same as keyring2
and `encoded_nonce != encoded_nonce2`

File: components/brave_wallet/browser/keyring_service.cc [hunk 5]
```diff
@@ -656,7 +670,10 @@ void KeyringService::CreateWallet(const std::string& password,
                                   CreateWalletCallback callback) {
   auto* keyring = CreateKeyring(mojom::kDefaultKeyringId, password);
   if (keyring) {
-    AddAccountForDefaultKeyring(GetAccountName(1));
+    AddAccountForKeyring(mojom::kDefaultKeyringId, GetAccountName(1));
+  }
+  if (IsFilecoinEnabled()) {
+    CreateKeyring(mojom::kFilecoinKeyringId, password);
```
Comment (darkdh) [inline components/brave_wallet/browser/keyring_service.cc:676]: This should be removed since we are doing lazy creation

File: components/brave_wallet/browser/brave_wallet_utils.cc [hunk 6]
```diff
@@ -135,6 +136,17 @@ const brave_wallet::mojom::EthereumChain kKnownNetworks[] = {
...
+    {brave_wallet::mojom::kFilecoinLocalhostChainId,
+     "Filecoin Ganache",
+     {kGanacheFilecoinLocalhostURL},
+     {},
+     {kGanacheFilecoinLocalhostURL},
+     "FIL",
+     "Filecoin",
+     18,
+     false}};
+
```
Comment (darkdh) [inline components/brave_wallet/browser/brave_wallet_utils.cc:149]: Please remove these unrelated changes from this file

File: browser/brave_wallet/keyring_service_unittest.cc [hunk 7]
```diff
@@ -2285,6 +2290,10 @@ TEST_F(KeyringServiceUnitTest, GetChecksumEthAddress) {
 }
 
 TEST_F(KeyringServiceUnitTest, UnknownKeyring) {
```
Comment (darkdh) [inline browser/brave_wallet/keyring_service_unittest.cc:2292]: please remove `service.CreateKeyring("one", "brave");` in the test and rename this case to `LazilyCreateKeyring`

File: components/brave_wallet/browser/keyring_service.cc (hunk 1 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc]: done

File: components/brave_wallet/browser/keyring_service.cc (hunk 2 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc]: done

File: components/brave_wallet/browser/keyring_service.cc (hunk 3 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc]: done

File: components/brave_wallet/browser/brave_wallet_utils.cc (hunk 6 above)
Comment (spylogsster) [reply components/brave_wallet/browser/brave_wallet_utils.cc]: done

File: browser/brave_wallet/keyring_service_unittest.cc (hunk 7 above)
Comment (spylogsster) [reply browser/brave_wallet/keyring_service_unittest.cc]: done

File: components/brave_wallet/browser/keyring_service.cc (hunk 5 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc]: done

File: components/brave_wallet/browser/filecoin_keyring.cc [hunk 8]
```diff
@@ -80,7 +80,7 @@ std::string FilecoinKeyring::ImportFilecoinSECP256K1Account(
   auto payload = BlakeHash(uncompressed_public_key, 20);
   int protocol = static_cast<int>(mojom::FilecoinAddressProtocol::SECP256K1);
   std::string address = network + std::to_string(protocol) +
-                        GetAddressInternal(payload, protocol);
+                        CreateAddressWithProtocol(payload, protocol);
```
Comment (darkdh) [inline components/brave_wallet/browser/filecoin_keyring.cc:83]: you should just call `GetAddressInternal` here

File: components/brave_wallet/browser/keyring_service.cc [hunk 9]
```diff
@@ -1292,10 +1342,16 @@ bool KeyringService::IsHardwareAccount(const std::string& account) const {
...
                             KeyringService::UnlockCallback callback) {
   if (!ResumeKeyring(mojom::kDefaultKeyringId, password)) {
-    encryptor_.reset();
+    encryptors_.erase(mojom::kDefaultKeyringId);
     std::move(callback).Run(false);
     return;
   }
+  if (IsFilecoinEnabled()) {
+    if (!ResumeKeyring(mojom::kFilecoinKeyringId, password)) {
+      VLOG(1) << __func__ << " Unable to unlock filecoin keyring";
```
Comment (darkdh) [inline components/brave_wallet/browser/keyring_service.cc:1351]: when unlocking failed, we should erase filecoin's encryptor

File: components/brave_wallet/browser/filecoin_keyring.cc (hunk 8 above)
Comment (spylogsster) [reply components/brave_wallet/browser/filecoin_keyring.cc:83]: `CreateAddressWithProtocol` filecoin specific and  `GetAddressInternal` calls it, there is additional protocol parameter

File: components/brave_wallet/browser/keyring_service.cc (hunk 9 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc:1352]: done

File: components/brave_wallet/browser/filecoin_keyring.cc (hunk 8 above)
Comment (darkdh) [reply components/brave_wallet/browser/filecoin_keyring.cc:83]: I meant these codes are duplicate
```
auto uncompressed_public_key = hd_key->GetUncompressedPublicKey();
  auto payload = BlakeHash(uncompressed_public_key, 20);
  int protocol = static_cast<int>(mojom::FilecoinAddressProtocol::SECP256K1);
  std::string address = network + std::to_string(protocol) +
                        CreateAddressWithProtocol(payload, protocol);
  ```
when you manage to get the network from settings, call `GetAddressInternal` to eliminate the duplicate

File: components/brave_wallet/browser/keyring_service.cc [hunk 10]
```diff
@@ -413,7 +422,12 @@ std::string KeyringService::GetAccountAddressForKeyring(
 }
 
 // static
-std::string KeyringService::GetAccountPathByIndex(size_t index) {
+std::string KeyringService::GetAccountPathByIndex(
+    size_t index,
+    const std::string& keyring_id) {
+  if (keyring_id == mojom::kFilecoinKeyringId) {
+    return base::NumberToString(index);
+  }
```
Comment (darkdh) [inline components/brave_wallet/browser/keyring_service.cc:430]: This is not right. Its derivation path should be `m/44'/461'/0'/0/index`
https://github.com/glifio/wallet/blob/3a36db74ce2d2ea55add1867d47ee1a13d75e3d7/utils/createPath/index.js#L6
you are currently using ETH's derivation path for FIL
`keyring->ConstructRootHDKey(*seed, kRootPath);` which is `m/44'/60'/0'/0` 
so keys you derived is not right, you can compare the addresses with filecoin wallet using the same mnemonics
they will be different

**Review by darkdh — CHANGES_REQUESTED** [review]
Derivation path for Filecoin keyring is wrong.
Check `Keyring::type` and give filecoin proper derivation path `m/44'/461'/0'/0` when calling 
`ConstructRootHDKey` in `KeyringService::CreateKeyringInternal`

File: components/brave_wallet/browser/keyring_service.cc (hunk 10 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc]: done, thanks for checking

File: browser/brave_wallet/keyring_service_unittest.cc (hunk 4 above)
Comment (spylogsster) [reply browser/brave_wallet/keyring_service_unittest.cc:482]: done

**Comment by spylogsster** [comment]
> We also need to test the same mnemonic is stored as different encrypted value in different keyrings. And test Lock and Unlock with multiple keyrings created

done

File: components/brave_wallet/browser/filecoin_keyring.cc (hunk 8 above)
Comment (spylogsster) [reply components/brave_wallet/browser/filecoin_keyring.cc:83]: there is a difference in network parameters, `GetAddressInternal` will create address using current network in wallet, but in import function that can be another imported address, the network is taken from parameters.

File: components/brave_wallet/browser/keyring_service.cc (hunk 9 above)
Comment (spylogsster) [reply components/brave_wallet/browser/keyring_service.cc]: done

**Review by darkdh — APPROVED** [review]
lgtm now
